lib/src/executable/watch.dart

```diff
@@ -298,15 +299,22 @@ class _Watcher {
           Uri newCanonicalUrl;
           try {
             newCanonicalUrl = _graph.importCache
-                .canonicalize(url, node.importer, node.canonicalUrl)
+                .canonicalize(url,
+                    baseImporter: node.importer,
+                    baseUrl: node.canonicalUrl,
+                    forImport: forImport)
                 ?.item2;
           } catch (_) {
-            // If the call to canonicalize failed, do nothing. We'll surface the
-            // error more nicely when we try to recompile the file.
+            // If the call to canonicalize failed, do nothing. We'll surface
+            // the error more nicely when we try to recompile the file.
           }
-          importChanged = newCanonicalUrl != node.upstream[url]?.canonicalUrl;
+          importChanged = newCanonicalUrl != node?.canonicalUrl;
         }
       }
+
+      node.upstream.forEach(recanonicalize);
+      forImport = true;
```

Comment:
Rather than using this as implicit state, I'd just pass it to `recanonicalize()`. It does mean you'll have to expand those `forEach()` calls, but I think the clarity of flow is worth it.